postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresConnector.java

```diff
@@ -0,0 +1,108 @@
+import io.cdap.cdap.etl.api.connector.ConnectorSpecRequest;
+import io.cdap.cdap.etl.api.connector.PluginSpec;
+import io.cdap.plugin.common.db.DBConnectorPath;
+import io.cdap.plugin.db.SchemaReader;
+import io.cdap.plugin.db.connector.AbstractDBSpecificConnector;
+import io.cdap.plugin.db.connector.AbstractDBSpecificConnectorConfig;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.mapreduce.lib.db.DBWritable;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * A Postgre SQL Database Connector that connects to Postgre SQL database via JDBC.
+ */
+@Plugin(type = Connector.PLUGIN_TYPE)
+@Name(PostgresConnector.NAME)
+@Description("This connector creates connections to a Postgre SQL database.")
+@Category("Database")
+public class PostgresConnector extends AbstractDBSpecificConnector<PostgresDBRecord> {
+  public static final String NAME = "Postgres";
+
+  protected PostgresConnector(AbstractDBSpecificConnectorConfig config) {
+    super(config);
+  }
+
+  @Override
+  public boolean supportSchema() {
+    return true;
+  }
+
+  @Override
+  protected Class<? extends DBWritable> getDBRecordType() {
+    return PostgresDBRecord.class;
+  }
+
+  protected void setConnectorSpec(ConnectorSpecRequest request, DBConnectorPath path,
+                                  ConnectorSpec.Builder builder) {
+    Map<String, String> properties = new HashMap<>();
+    properties.put(PostgresSource.PostgresSourceConfig.NAME_USE_CONNECTION, "true");
+    properties.put(PostgresSource.PostgresSourceConfig.NAME_CONNECTION, request.getConnectionWithMacro());
+    builder.addRelatedPlugin(new PluginSpec(PostgresConstants.PLUGIN_NAME, BatchSource.PLUGIN_TYPE, properties));
+
+    String table = path.getTable();
+    if (table == null) {
+      return;
+    }
+
+    properties.put(PostgresSource.PostgresSourceConfig.IMPORT_QUERY,
+                   String.format("SELECT * FROM %s.%s.%s;", path.getDatabase(), path.getSchema(), table));
+    properties.put(PostgresSource.PostgresSourceConfig.NUM_SPLITS, "1");
+  }
+
+  @Override
+  protected SchemaReader getSchemaReader() {
+    return new PostgresSchemaReader();
+  }
+
+  @Override
+  public StructuredRecord transform(LongWritable longWritable, PostgresDBRecord record) {
+    return record.getRecord();
+  }
+
+  protected ResultSet queryDatabases(Connection connection) throws SQLException {
+    return connection.createStatement().executeQuery(String.format(
+      "SELECT datname AS %s FROM pg_database WHERE datistemplate = false;",
+      RESULTSET_COLUMN_TABLE_CAT));
+  }
+
+  @Override
+  protected String getConnectionString(@Nullable String database) {
+    if (database == null) {
+      return config.getConnectionString();
+    }
+    return String
+      .format(PostgresConstants.POSTGRES_CONNECTION_STRING_FORMAT, config.getHost(), config.getPort(), database);
```

Comment:
Why do we have `POSTGRES_CONNECTION_STRING_FORMAT` and `POSTGRE_CONNECTION_STRING_FORMAT` in `PostgresConnectorConfig`?

Reply:
Good question. Because initially for a connector to test connection, we don't know what database to connect to (User dont' need to specify a database) , so the format in Connector is without database. Only when user dive into a specific database, we need to reconstruct the connection string for that database and here the format contains database.

Reply:
renamed them respectively to make it more clear.